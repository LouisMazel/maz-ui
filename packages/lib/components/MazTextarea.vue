<script lang="ts">
/* eslint-disable import/first */
import type { Color } from './types'

export type { Color }

export interface MazTextareaProps<T extends string | undefined | null> {
  /** Style attribut of the component root element */
  style?: HTMLAttributes['style']
  /** Class attribut of the component root element */
  class?: HTMLAttributes['class']
  /** @model The value of the textarea */
  modelValue?: T
  /** The id of the textarea */
  id?: string
  /** The name of the textarea */
  name?: string
  /** The label of the textarea */
  label?: string
  /** The placeholder of the textarea */
  placeholder?: string
  /** If the textarea is required */
  required?: boolean
  /** If the textarea is disabled */
  disabled?: boolean
  /** If the textarea is readonly */
  readonly?: boolean
  /** If the textarea has an error */
  error?: boolean
  /** If the textarea has a success */
  success?: boolean
  /** If the textarea has a warning */
  warning?: boolean
  /** The hint of the textarea */
  hint?: string
  /** The color of the textarea */
  color?: Color
  /**
   * Size radius of the component's border
   * @values `'none' | 'sm' | 'md' | 'lg' | 'xl' | 'full'`
   */
  roundedSize?: 'none' | 'sm' | 'md' | 'lg' | 'xl' | 'full'
  /**
   * If the textarea has a padding
   * @default true
   */
  padding?: boolean
  /**
   * If the textarea has a transparent background
   * @default false
   */
  transparent?: boolean
  /**
   * If the textarea has no border
   * @default false
   */
  border?: boolean
  /**
   * The alignment of the append slot
   * @values `'start' | 'end' | 'center' | 'space-between' | 'space-around' | 'space-evenly'`
   * @default 'end'
   */
  appendJustify?: 'start' | 'end' | 'center' | 'space-between' | 'space-around' | 'space-evenly'
}
</script>

<script lang="ts" setup generic="T extends string | undefined | null">
import { computed, type HTMLAttributes, onBeforeUnmount, onMounted, ref, useSlots } from 'vue'
import { useInstanceUniqId } from '../modules/composables/useInstanceUniqId'
import { TextareaAutogrow } from './MazTextarea/textarea-autogrow'

defineOptions({
  inheritAttrs: false,
})

const props = withDefaults(defineProps<MazTextareaProps<T>>(), {
  style: undefined,
  class: undefined,
  modelValue: undefined,
  id: undefined,
  name: 'MazTextarea',
  label: undefined,
  placeholder: undefined,
  required: false,
  disabled: false,
  readonly: false,
  error: false,
  success: false,
  warning: false,
  hint: undefined,
  color: 'primary',
  padding: true,
  transparent: false,
  border: true,
  appendJustify: 'end',
})

const emits = defineEmits<{
  /**
   * Emitted when the value of the textarea change
   * @property {string | undefined} value - The value of the textarea
   */
  (event: 'update:model-value', value?: T): void
  /**
   * Emitted when the value of the textarea change
   * @property {string | undefined} value - The value of the textarea
   */
  (event: 'input', value?: T): void
  /**
   * Emitted when the textarea is focused
   * @property {Event} value - The event
   */
  (event: 'focus', value: FocusEvent): void
  /**
   * Emitted when the textarea is blurred
   * @property {Event} value - The event
   */
  (event: 'blur', value: FocusEvent): void
  /**
   * Emitted when the textarea value change
   * @property {Event} value - The event
   */
  (event: 'change', value: Event): void
}>()

let textareaAutogrow: TextareaAutogrow | undefined

const instanceId = useInstanceUniqId({
  componentName: 'MazTextarea',
  providedId: props.id,
})

const TextareaElement = ref<HTMLTextAreaElement>()
const isFocused = ref(false)
const hasValue = computed(() => props.modelValue !== undefined && props.modelValue !== '')

onMounted(() => {
  if (TextareaElement.value) {
    textareaAutogrow = new TextareaAutogrow(TextareaElement.value)
  }
})

onBeforeUnmount(() => {
  textareaAutogrow?.disconnect()
})

const inputValue = computed({
  get: () => props.modelValue,
  set: (value) => {
    emits('update:model-value', value)
    emits('input', value)
  },
})

function focus(event: FocusEvent) {
  emits('focus', event)
  isFocused.value = true
}

function blur(event: FocusEvent) {
  emits('blur', event)
  isFocused.value = false
}

function change(event: Event) {
  emits('change', event)
}

const slots = useSlots()

const hasLabelOrHint = computed(() => props.label || props.hint || !!slots.label)

const shouldUp = computed(
  () => hasLabelOrHint.value && (isFocused.value || hasValue.value || !!props.placeholder),
)

const hasAppend = computed(() => !!slots.append)

// eslint-disable-next-line sonarjs/cognitive-complexity
const borderStyle = computed(() => {
  if (props.error)
    return 'maz-border-danger'
  if (props.success)
    return 'maz-border-success'
  if (props.warning)
    return 'maz-border-warning'
  if (isFocused.value) {
    if (props.color === 'black')
      return 'maz-border-black'
    if (props.color === 'danger')
      return 'maz-border-danger'
    if (props.color === 'info')
      return 'maz-border-info'
    if (props.color === 'primary')
      return 'maz-border-primary'
    if (props.color === 'secondary')
      return 'maz-border-secondary'
    if (props.color === 'success')
      return 'maz-border-success'
    if (props.color === 'warning')
      return 'maz-border-warning'
    if (props.color === 'white')
      return 'maz-border-white'
  }
  return '--default-border'
})

const hasBorderStyle = computed(() => borderStyle.value !== '--default-border')
</script>

<template>
  <label
    class="m-textarea"
    :for="instanceId"
    :class="[
      {
        '--is-disabled': disabled,
        '--has-label': hasLabelOrHint,
        '--background-transparent': transparent,
        '--padding': padding,
        '--border': border,
        '--has-border-style': hasBorderStyle,
      },
      borderStyle,
      roundedSize ? `--rounded-${roundedSize}` : '--rounded',
      props.class,
    ]"
    :style="[style, `--append-justify: ${appendJustify}`]"
  >
    <!-- eslint-disable vuejs-accessibility/label-has-for -->
    <label
      v-if="hasLabelOrHint"
      :for="instanceId"
      class="m-textarea__label"
      :class="[
        {
          'maz-text-danger-600': error,
          'maz-text-success-600': success,
          'maz-text-warning-600': warning,
          '--has-state': error || warning || success,
          '--should-up': shouldUp,
        },
      ]"
    >
      <!-- @slot Label - Replace the label -->
      <slot v-if="!hint" name="label">
        {{ label }}
      </slot>
      <span v-else>
        {{ hint }}
      </span>
      <sup v-if="required">*</sup>
    </label>
    <!-- eslint-enable vuejs-accessibility/label-has-for -->

    <textarea
      :id="instanceId"
      ref="TextareaElement"
      v-bind="$attrs"
      v-model="inputValue"
      :placeholder
      :name
      :disabled
      :required
      :class="{ '--has-append': hasAppend }"
      v-on="{
        blur,
        focus,
        change,
      }"
    />
    <div v-if="hasAppend" class="m-textarea__append">
      <!-- @slot Append - Replace the append -->
      <slot name="append" />
    </div>
  </label>
</template>

<style lang="postcss" scoped>
  .m-textarea {
  @apply maz-min-h-[6.25rem] maz-relative maz-flex maz-flex-col maz-align-top maz-text-normal;

  &:not(.--background-transparent) {
    @apply maz-bg-color dark:maz-bg-color-light;
  }

  &.--border {
    @apply maz-border maz-border-solid;

    &:not(.--has-border-style) {
      @apply maz-border-border dark:maz-border-color-lighter;
    }
  }

  &.--padding {
    @apply maz-px-4 maz-py-3;
  }

  &.--rounded-sm {
    @apply maz-rounded-sm;
  }

  &.--rounded-md {
    @apply maz-rounded-md;
  }

  &.--rounded-lg {
    @apply maz-rounded-lg;
  }

  &.--rounded-xl {
    @apply maz-rounded-xl;
  }

  &.--rounded-full {
    @apply maz-rounded-full;
  }

  &.--rounded {
    @apply maz-rounded;
  }

  &.--has-label {
    @apply maz-pt-6;
  }

  &__append {
    @apply maz-inline-flex;

    justify-content: var(--append-justify);
  }

  textarea {
    @apply maz-w-full maz-resize-y maz-outline-none maz-bg-transparent;

    &.--has-append {
      @apply maz-pb-4;
    }
  }

  &.--is-disabled {
    textarea {
      @apply maz-cursor-not-allowed maz-border-border maz-bg-color-lighter maz-text-muted dark:maz-border-color-lighter;
    }
  }

  &__label {
    @apply maz-pointer-events-none maz-absolute maz-block maz-w-max maz-origin-top-left maz-truncate;
    @apply maz-left-4 maz-top-3 maz-leading-6;
    @apply maz-flex maz-flex-center;

    transition: transform 200ms cubic-bezier(0, 0, 0.2, 1) 0ms;

    &.--should-up {
      transform: scale(0.8) translateY(-0.65rem);
    }

    &:not(.--has-state) {
      @apply maz-text-muted;
    }
  }
}
</style>
