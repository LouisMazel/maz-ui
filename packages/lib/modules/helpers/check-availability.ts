export function checkAvailability<T>(getRef: () => T | null | undefined, callback: (component: NonNullable<T>) => void, options?: { maxAttempts?: number, interval?: number, errorMessage?: string }) {
  const opts = {
    maxAttempts: 20,
    interval: 100,
    ...options,
  }

  let attempts = 0

  function check() {
    const ref = getRef()
    if (ref !== null && ref !== undefined) {
      callback(ref as NonNullable<T>)
    }
    else if (attempts < opts.maxAttempts) {
      attempts++
      setTimeout(check, opts.interval)
    }
    else {
      console.warn(opts.errorMessage || `[maz-ui](checkAvailability) Nothing found after ${opts.maxAttempts} attempts for ref ${ref}`)
    }
  }

  check()
}
