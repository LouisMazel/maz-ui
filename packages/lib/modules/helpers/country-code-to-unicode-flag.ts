import { truthyFilter } from './truthy-filter'

export function countryCodeToUnicodeFlag(locale: string) {
  return [...locale]
    .map((letter) => {
      const code = letter.codePointAt(0)
      return code ? (code % 32) + 0x1_f1_e5 : undefined
    })
    .filter(truthyFilter)
    .map((n) => String.fromCodePoint(n))
    .join('')
}
