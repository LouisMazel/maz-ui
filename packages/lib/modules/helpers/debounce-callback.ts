let timeout: NodeJS.Timeout | null = null

export function debounceCallback(callback: (...args: unknown[]) => unknown, delay: number) {
  if (timeout) {
    clearTimeout(timeout)
  }

  timeout = setTimeout(callback, delay)
}
