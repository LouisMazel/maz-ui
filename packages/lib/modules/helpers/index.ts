export * from './inject-strict'
export * from './debounce'
export * from './truthy-filter'
export * from './mount-component'
export * from './idle-timeout'
export * from './user-visibility'
export * from './sleep'
export * from './is-client'
export * from './script-loader'
export * from './country-code-to-unicode-flag'
export * from './swipe-handler'
export * from './throttle'
