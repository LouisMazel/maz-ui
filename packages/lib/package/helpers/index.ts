export { injectStrict } from './inject-strict'
export { debounce } from './debounce'
export { truthyFilter } from './truhty-filter'
export { mount } from './mount-component'
