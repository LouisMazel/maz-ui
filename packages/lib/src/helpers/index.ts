export * from './checkAvailability'
export * from './countryCodeToUnicodeFlag'
export * from './countryFlagFromFlagCdn'
export * from './debounce'
export * from './debounceCallback'
export * from './debounceId'
export * from './freezeValue'
export * from './idleTimeout'
export * from './injectStrict'
export * from './isClient'
export * from './isEqual'
export * from './isStandaloneMode'
export * from './mountComponent'
export * from './normalizeString'
export * from './scriptLoader'
export * from './sleep'
export * from './swipeHandler'
export * from './throttle'
export * from './throttleId'
export * from './truthyFilter'
export * from './userVisibility'
