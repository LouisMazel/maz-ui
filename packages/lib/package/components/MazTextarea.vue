<template>
  <div
    class="m-textarea"
    :class="{
      '--is-disabled': disabled,
      '--has-label': hasLabelOrHint,
    }"
  >
    <label
      v-if="hasLabelOrHint"
      :for="instanceId"
      class="m-textarea__label"
      :class="[
        {
          'maz-text-danger-600': error,
          'maz-text-success-600': success,
          'maz-text-warning-600': warning,
          '--has-state': error || warning || success,
          '--should-up': shouldUp,
        },
      ]"
    >
      {{ hint || label }}
      <sup v-if="required">*</sup>
    </label>

    <!-- eslint-disable vue/no-deprecated-html-element-is -->
    <textarea
      :id="instanceId"
      ref="TextareaElement"
      v-bind="$attrs"
      v-model="inputValue"
      :name="name"
      :disabled="disabled"
      :required="required"
      :class="[borderStyle]"
      v-on="{
        blur,
        focus,
        change,
      }"
    />
    <!-- eslint-enable vue/no-deprecated-html-element-is -->
  </div>
</template>

<script lang="ts">
  export type { Color } from './types'
</script>

<script lang="ts" setup>
  import { useInstanceUniqId } from '@package/helpers'
  import { computed, onBeforeUnmount, onMounted, ref, type PropType } from 'vue'
  import { getCurrentInstance } from 'vue'
  import { TextareaAutogrow } from './MazTextarea/textarea-autogrow'
  import type { Color } from './types'

  const props = defineProps({
    modelValue: {
      type: String,
      default: undefined,
    },
    id: { type: String, default: undefined },
    name: { type: String, default: 'MazTextarea' },
    label: { type: String, default: undefined },
    required: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    readonly: { type: Boolean, default: false },
    error: { type: Boolean, default: false },
    success: { type: Boolean, default: false },
    warning: { type: Boolean, default: false },
    hint: { type: String, default: undefined },
    color: {
      type: String as PropType<Color>,
      default: 'primary',
      validator: (value: string) => {
        return [
          'primary',
          'secondary',
          'warning',
          'danger',
          'info',
          'success',
          'white',
          'black',
          'transparent',
        ].includes(value)
      },
    },
  })

  const emits = defineEmits(['input', 'focus', 'blur', 'change'])

  const instance = getCurrentInstance()

  let textareaAutogrow: TextareaAutogrow | undefined

  const { instanceId } = useInstanceUniqId({
    componentName: 'MazTextarea',
    instance,
    providedId: props.id,
  })

  const TextareaElement = ref<HTMLTextAreaElement>()
  const isFocused = ref(false)
  const hasValue = computed(
    () => props.modelValue !== undefined && props.modelValue !== '',
  )

  onMounted(() => {
    if (TextareaElement.value) {
      textareaAutogrow = new TextareaAutogrow(TextareaElement.value)
    }
  })

  onBeforeUnmount(() => {
    textareaAutogrow?.disconnect()
  })

  const inputValue = computed({
    get: () => props.modelValue,
    set: (value: unknown) => emits('input', value),
  })

  const focus = (event: Event) => {
    emits('focus', event)
    isFocused.value = true
  }

  const blur = (event: Event) => {
    emits('blur', event)
    isFocused.value = false
  }

  const change = (event: Event) => emits('change', event)

  const hasLabelOrHint = computed(() => props.label || props.hint)

  const shouldUp = computed(
    () => hasLabelOrHint.value && (isFocused.value || hasValue.value),
  )

  const borderStyle = computed(() => {
    if (props.error) return 'maz-border-danger'
    if (props.success) return 'maz-border-success'
    if (props.warning) return 'maz-border-warning'
    if (isFocused.value) {
      if (props.color === 'black') return 'maz-border-black'
      if (props.color === 'danger') return 'maz-border-danger'
      if (props.color === 'info') return 'maz-border-info'
      if (props.color === 'primary') return 'maz-border-primary'
      if (props.color === 'secondary') return 'maz-border-secondary'
      if (props.color === 'success') return 'maz-border-success'
      if (props.color === 'warning') return 'maz-border-warning'
      if (props.color === 'white') return 'maz-border-white'
    }
    return '--default-border'
  })
</script>

<style lang="postcss" scoped>
  .m-textarea {
    @apply maz-relative maz-flex maz-flex-col;

    textarea {
      @apply maz-min-h-[6.25rem] maz-w-full maz-resize-y maz-rounded maz-border
        maz-bg-color maz-p-4 maz-text-normal maz-outline-none;

      &.--default-border {
        @apply maz-border-gray-200;
      }
    }

    &.--has-label {
      textarea {
        @apply maz-pt-6;
      }
    }

    &.--is-disabled {
      textarea {
        @apply maz-cursor-not-allowed maz-border-gray-200 maz-bg-color-lighter maz-text-muted;
      }
    }

    &__label {
      @apply maz-pointer-events-none maz-absolute maz-block maz-w-max maz-origin-top-left maz-truncate;
      @apply maz-left-4 maz-top-3 maz-leading-6;
      @apply maz-flex maz-flex-center;

      transition: transform 200ms cubic-bezier(0, 0, 0.2, 1) 0ms;

      &.--should-up {
        transform: scale(0.8) translateY(-0.65rem);
      }

      &:not(.--has-state) {
        @apply maz-text-muted;
      }
    }
  }

  html.dark {
    & .m-textarea {
      textarea {
        @apply maz-bg-color-light;

        &.--default-border {
          @apply maz-border-color-lighter;
        }
      }

      &.--is-disabled {
        textarea {
          @apply maz-border-color-lighter;
        }
      }
    }
  }
</style>
