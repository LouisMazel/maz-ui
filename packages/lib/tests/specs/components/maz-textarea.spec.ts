import { type VueWrapper, shallowMount } from '@vue/test-utils'
import MazTextarea from '@components/MazTextarea.vue'
import { TextareaAutogrow } from '@components/MazTextarea/textarea-autogrow'
import type { ComponentPublicInstance } from 'vue'
import { elementEmitEvent } from '@tests/helpers/document-event'

describe('components/MazTextarea/textarea-autogrow.ts', () => {
  let textareaElement: HTMLTextAreaElement

  beforeEach(() => {
    textareaElement = document.createElement('textarea')
    const _textareaAutogrow = new TextareaAutogrow(textareaElement)
  })

  test('Should add css style on init', () => {
    /* eslint-disable @typescript-eslint/ban-ts-comment */
    // @ts-ignore
    expect(textareaElement.style.resize).toBe('none')
    // @ts-ignore
    expect(textareaElement.style.boxSizing).toBe('border-box')
    /* eslint-enable @typescript-eslint/ban-ts-comment */
  })

  test('Should add css style on focus event', () => {
    elementEmitEvent(textareaElement, 'focus')

    expect(textareaElement.style.height).toBe('0px')
    expect(textareaElement.style.overflow).toBe('hidden')
  })

  test('Should add css style on resize event', () => {
    elementEmitEvent(textareaElement, 'focus')
    elementEmitEvent(window, 'resize')

    expect(textareaElement.style.height).toBe('0px')
    expect(textareaElement.style.overflow).toBe('hidden')
  })
})

describe('components/MazTextarea.vue', () => {
  expect(MazTextarea).toBeTruthy()

  const wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }> = shallowMount(
    MazTextarea,
    {
      props: {
        modelValue: 'Un text',
      },
    },
  )

  test('Should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Should have the model value', () => {
    expect(wrapper.vm.modelValue).toBe('Un text')
  })

  test('Should have an uniq id', () => {
    expect(wrapper.vm.instanceId).toBe('MazTextarea-1')
  })

  test('Should up the label if the component is focused', async () => {
    await wrapper.setProps({
      modelValue: undefined,
      label: 'Label',
    })
    wrapper.vm.focus()
    expect(wrapper.vm.shouldUp).toBeTruthy()
  })

  test('Should emit focus event', () => {
    wrapper.vm.focus()
    expect(wrapper.emitted().focus).toBeTruthy()
  })

  test('Should emit blur event', () => {
    wrapper.vm.blur()
    expect(wrapper.emitted().blur).toBeTruthy()
  })

  test('Should emit change event', () => {
    wrapper.vm.change()
    expect(wrapper.emitted().change).toBeTruthy()
  })

  test('Should have border color according with the color prop', async () => {
    await wrapper.setProps({
      error: true,
    })
    expect(wrapper.vm.borderStyle).toBe('maz-border-danger')
    await wrapper.setProps({
      error: false,
      warning: true,
    })
    expect(wrapper.vm.borderStyle).toBe('maz-border-warning')
    await wrapper.setProps({
      error: false,
      warning: false,
      success: true,
    })
    expect(wrapper.vm.borderStyle).toBe('maz-border-success')
    await wrapper.setProps({
      error: false,
      warning: false,
      success: false,
    })
    wrapper.vm.focus()
    expect(wrapper.vm.borderStyle).toBe('maz-border-primary')
    await wrapper.setProps({
      color: 'secondary',
    })
    expect(wrapper.vm.borderStyle).toBe('maz-border-secondary')
  })
})
