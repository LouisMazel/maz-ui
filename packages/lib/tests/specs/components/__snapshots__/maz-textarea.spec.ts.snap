// Vitest Snapshot v1

exports[`components/MazTextarea.vue > Should match with the snapshot 1`] = `
"<div class=\\"m-textarea\\" data-v-144c4295=\\"\\">
  <!--v-if--><textarea id=\\"MazTextarea-1\\" name=\\"MazTextarea\\" class=\\"--default-border\\" data-v-144c4295=\\"\\" style=\\"resize: none; box-sizing: border-box;\\"></textarea>
</div>"
`;
